File: algorithm/token_bucket.go
package algorithm

import (
	"math"
	"sync"
	"time"
)

type TokenBucketState struct {
	mu         sync.Mutex
	tokens     float64
	lastUpdate time.Time
	expiry     time.Time
}

type TokenBucketConfig struct {
	Capacity float64

	RefillRate float64
}

func (cfg *TokenBucketConfig) Evaluate(state *TokenBucketState, cost int, now time.Time, window time.Duration) Result {
	state.mu.Lock()
	defer state.mu.Unlock()

	if state.lastUpdate.IsZero() {
		state.tokens = cfg.Capacity
		state.lastUpdate = now
	}

	elapsed := now.Sub(state.lastUpdate).Seconds()
	if elapsed < 0 {
		elapsed = 0
	}
	refilled := elapsed * cfg.RefillRate
	state.tokens = math.Min(cfg.Capacity, state.tokens+refilled)
	state.lastUpdate = now
	state.expiry = now.Add(window * 2)

	fCost := float64(cost)
	if state.tokens >= fCost {
		state.tokens -= fCost
		return Result{
			Allowed:   true,
			Remaining: int(math.Floor(state.tokens)),
		}
	}

	deficit := fCost - state.tokens
	waitSeconds := deficit / cfg.RefillRate
	return Result{
		Allowed:    false,
		Remaining:  0,
		RetryAfter: time.Duration(math.Ceil(waitSeconds)) * time.Second,
	}
}

func (s *TokenBucketState) IsExpired(now time.Time) bool {
	s.mu.Lock()
	defer s.mu.Unlock()
	return !s.expiry.IsZero() && now.After(s.expiry)
}

File: backend/lua/token_bucket.lua
-- KEYS[1]   : rate limit key
-- ARGV[1]   : capacity (max tokens)
-- ARGV[2]   : refill rate (tokens per second)
-- ARGV[3]   : current timestamp (unix ms)
-- ARGV[4]   : cost (tokens to consume)
-- ARGV[5]   : TTL in seconds

local key         = KEYS[1]
local capacity    = tonumber(ARGV[1])
local refill_rate = tonumber(ARGV[2])
local now         = tonumber(ARGV[3])
local cost        = tonumber(ARGV[4])
local ttl         = tonumber(ARGV[5])

local state       = redis.call("HMGET", key, "tokens", "last_update")
local tokens      = tonumber(state[1])
local last_update = tonumber(state[2])

if tokens == nil then
    tokens      = capacity
    last_update = now
end

local elapsed    = math.max(0, now - last_update) / 1000.0
local new_tokens = math.min(capacity, tokens + elapsed * refill_rate)

redis.call("HMSET", key, "tokens", new_tokens, "last_update", now)
redis.call("EXPIRE", key, ttl)

if new_tokens >= cost then
    new_tokens = new_tokens - cost
    redis.call("HMSET", key, "tokens", new_tokens, "last_update", now)
    return {1, math.floor(new_tokens), 0}
else
    local deficit      = cost - new_tokens
    local retry_after  = math.ceil(deficit / refill_rate)
    return {0, math.floor(new_tokens), retry_after}
end
